### ISSUE TITLE
Open with button to only show apps a user has selected

### ISSUE THREAD

**Issue by dtarb** [hydroshare/hydroshare#1161]
To avoid the random appearance of any app any user has made public in HydroShare (that was a problem e.g. #1132) establish the following model for users to "activate" apps that they want to use.  Only apps a user has "activated" show up on the open with button presented to a user.  
- The "Activate" feature for app resources should be similar to the "Favorite" feature for any resource.  The system should maintain a list of app resources each user has activated (or a list of users who have activated each app resource - internal implementation suggested to be the same as Favorite).  
- Activating an app resource should add it to the users My Resources page.  The My Resources page should let a user activate and deactivate an app resource similar to the way a user can favorite and unfavorite a resource.  
- An app resource landing page should include a button (suggest top right with other buttons) for a user to "Activate".  
- This implementation would mean a user has to look at an app landing page and activate any app they want to have on their "Open with" button.  The abstract on an App landing page then serves to inform users selection based on what the App does, an important mechanism for communication of this functionality and education of users.
- All apps set for a resource type show up in the Web Apps tab of a resource landing page.  I suggest a heading similar to "Community apps you may choose to use with this resource" and an info link that goes to the app landing page and run link that launches the app.  The web apps tab thus serves as a way for users to discover relevant apps, and to run them if they wish without having to Activate (and without necessarily having to be logged in if the app permits this).

**Comment by horsburgh** [comment]
Another potentially much simpler option is to simply let anyone publish and share an app.  But we as the HydroShare team review them and put the best ones (and by "best" I mean working and doing something useful) in the list.

**Comment by dtarb** [comment]
I am trying to avoid a solution where "we" the hydroshare team have to moderate this as it limits scalability.  But as an initial solution, I would be happy to have an approved app list, especially if you are willing to be the moderator who reviews it.

**Comment by horsburgh** [comment]
If we had masses of app developers developing apps, I would worry more about scalability.  And, considering how much development time it would probably take to implement the system above - I think we have to consider other options, until scalability becomes an issue.

**Comment by hyi** [comment]
@shawncrawley I am assigning this to you initially. Feel free to reassign if needed.

**Comment by dtarb** [comment]
I am bumping this up in priority and assigning to BYU team @danames to prevent the rogue appearance of improper apps on all pages such as the "Trophic State HydroServer" that is really just a web page that is currently in HydroShare.

I suggest the following, 

The Open with menu should hold only:
1.	Apps that have been approved by the developer team.  This requires establishing functionality for an admin to check a setting on the app instance to enable this.
2.	Apps that a user has identified that they want to use by either adding them to “My Resources” or  “Favoriting” them.

This should be an additional layer of filtering that goes on top of whether the user has permissions on the app and whether the app is configured for the type of resource the user has open.  

Also, since we will shortly have resources with file types, the app instance settings need to be extended to be enabled when a resource has a file type that the app can handle.  In this case there may be merit in having Open with on the file right click menu in addition to at the resource level.

All apps that have been enabled based on resource or file type should still appear under the web apps tab where there should also be functionality to perform the action (add to my resources or favorite) that puts them on the "Open With" menu (and a few words that explain how to do this).

**Comment by dtarb** [comment]
@alvacouch please add your suggestions here.

**Comment by zhiyuli** [comment]
@danames @dtarb 

To simplify this a little bit. can we directly use the existing "Add to my resource" button? 

I am thinking the "Open With" dropdown should list the following webapp resources:
1) all webapp resources that appear on a user's "My Resource" page including
   a) webapps this user owns;
   b) webapps shared with this user; 
   c) webapps this user discovered and added to his "My Resource" page through "Add to my resource button"

2) webapps approved by system Admin: add a new admin-editable metadata field to control this.

**Comment by horsburgh** [comment]
@zhiyuli - This is getting better (narrower), but still leaves the possibility that apps can appear in the "Open With" list for a resource where the app has no clue how to actually open that resource.

Or is your condition all apps that meet conditions 1 and 2 AND know how to actually open the resource? Is there a setting or attribute on Apps that designates which resource/file types they can operate on?

**Comment by zhiyuli** [comment]
@horsburgh 
Currently the "file type" concept is not implemented in webapp yet. But we already have "supported resource type" concept implemented, so a webapp only shows up on resource landing pages of those supported types ("open with" dropdown and "webapp" tab page)

The above 2 conditions are to further limit which webapps should show up in the "open with" dropdown.

**Comment by horsburgh** [comment]
Oh - sweet. This seems reasonable to me and uses our existing model for users adding things to their resources, which I think is consistent with what @dtarb suggested.

**Comment by dtarb** [comment]
@zhiyuli indicated that this functionality has been implemented and demo's but not merged and deployed due to not having an easy way to mark a web app as Approved.

I suggest we use a HydroShare resource to hold a list of approved apps (e.g. as a CSV text file with their name and resource ID).  Then access to editing this resource can be managed through HydroShare access control.  Then Open With functionality can read this file to obtain approved apps.  The same file could be used to generate a simple web page for launching apps as described in #2301 instead of each app requiring a Tethys implementation to just provide a launching button which is cumbersome for non-Tethys apps.

**Comment by pkdash** [comment]
I think we implement a new page for managing apps and provide a menu option (MANAGE APPS) on the upper navigation bar to access this page. This page should list all web apps similar to listing of resources in my resources page (allow sorting). For each web app provide a checkbox for approval. Each row of web app should provide necessary information (title, purpose, web app author etc) about the web app to help in making the approval decision for the app. For each web app we should track who approved and the date of approval. Only users in the 'Manage Apps' group can have access to the MANAGE APPS menu. Since the approved status is maintained as part of the web app, then we can list only the approved apps for the specific resource type/file type/mime type when the Open With option is selected.

This obviously involves more work then what @dtarb has suggested above. However, I think in the long run this is the kind of functionality at a minimum we need for managing apps.

**Comment by zhiyuli** [comment]
@danames @dtarb @horsburgh 

@aphelionz has added a new "ShowOnOpenWithList" checkbox in his PR for “app replacement app”. Since the business logic of checkbox is unclear, we decided to remove that checkbox from that PR for now.

I then reminded @danames of a simplified workflow we discussed in April (see early conversations in this issue), which was to re-use existing “Add Favorite” button instead of adding any new checkbox.

The workflow is:
An app will show up in the Open-With-List on a resource landing page a user is viewing at, if this app supports this Resource Type, and meets **any** of the following 2 conditions:
1) This app resource is on this user's "My Resource Page" 
2) This app is public/published, and has been marked as a "approved app" by admin

An app will meet condition (1) if
a)this user is the Owner of this app, regardless of its sharing status  (published/public/discoverable/private) 
or,
b)this app has been explicitly shared with this user regardless of app’s sharing status  (published/public/discoverable/private)
or,
c)this app is public/published, and this user has discovered this app through discover page and added it as his favorite resource 

@danames  supports this approach. 
I would like to check with @dtarb and @horsburgh

Thanks

**Comment by dtarb** [comment]
I think that the idea suggested by @danames in #2336 and @aphelionz in #2322 of a specific button "Show on Open With" is better than the approach I initially suggested in #1161 of having show on open with behavior determined through being added to "My Resources" or "Favoriting".  I think that a check box explicity labeled by what it does will be clearer to users than having to do something labeled something else.  Therefore I suggest the following implementaiton and business logic (Note that this is somewhat similar to the Activate concept at the very beginning of this issue, but updated, simplified and repeated for completeness):

The system should maintain a list of app resources each user has selected for "Show on Open With" (or a list of users who have activated each app resource - internal implementation I leave to a developer to choose what is best).

An app resource landing page should include a check box "Show on Open With" for a user to select if they want that web app to be on their open with menus. 

This "Show on Open With" checkbox should be visible and able to be checked or unchecked for any web app resource that a user can open the landing page for in view mode.  Checking this box is not an edit action on a resource so does not require edit permissions, and should be possible in view mode.  It does not matter whether the resource is public or has been shared explicitly with a user.

Business logic for whether a web app should appear on "Open with" menu
1. Check that the user has permission to view the web app resource (it has to be public or they have to have explicit view permission).
2. If the web app resource has been CUAHSI approved it can appear on the "Open with" menu.  This takes precedence over whether or not a user has explicitly checked "Show on Open With".  If a web app resource is CUAHSI approved there is no evaluation of "Show on Open With", and whether or not show on open with is checked has no effect.
3. If point 2 is not met, but a user has explicitly checked "Show on Open With" it can appear on an "Open with" menu

Once the above 3 points have been evaluated the rules based on resource and file type as set for the web app resource instance (by the owner or user with edit permissions for the web app instance) need to be checked
1. Check whether the resource type matches the type for which the web app is configured.  This relies on existing web app metadata that lets a web app be configured to apply to specific resource types.
2. Check whether there are requirements for specific file types to be present in the resource and whether these requirements are met.  A web app may be configured to apply to specific file types within a multitype resource (such as a composite resource).  This configuration should indicate whether the web app should appear on the resource level open with menu (making the menu entry only appear if there is a file of the specified file type in the resource) or at the file type level (making the app appear on a file level open with menu for a specific file type). 

There also need to be resource app entries that specify App-launching URL patterns separately for both resource level and file level open with actions, because the action for a resource level open with may need to be different from the action for a file level open with. This point also links to additional metadata needed for a web app resource to support this functionality that is addressed in #1610 or #2336.  Additional metadata is needed to support this business logic, and app launching URL patterns.

Note that the above suggestions decouple the concept of "Show on Open With" from other resource concepts "Favorite" and "Add to my resources".  I think that this decoupling makes things simpler and avoids potential user confusion, and also means that implementation can ignore changing anything related to these other concepts.

Note also that I decided that using an app required view access to the full resource.  Access to only the metadata (as would be the case if it was only discoverable) would not be sufficient.  I could have gone the other way in this (and could easily be persuaded to change), as an app resource is really a metadata only resource so discoverable and public are equivalent.

**Comment by horsburgh** [comment]
Wow. This seems really complicated. How many CUAHSI approved apps are there going to be? How is a user supposed to know an App exists so they can go to its landing page and tick a box for that App to appear? Seems like all the complication of vetting apps should be done before they are CUAHSI approved. On the user end, it should just be as simple as choosing to include an app in the "Open with" list or not. 

Why not do that action right on the "Open with" button - e.g., click the button and there would be an option of "Edit this list". Then the user would get a pop up that allows them to tick boxes next to the apps that are CUAHSI approved and appropriate for the type of resource they are viewing. We would curate the list of approved Apps.

**Comment by zhiyuli** [comment]
@dtarb 
I have 2 question about this statement:
"Check whether there are requirements for specific file types to be present in the resource and whether these requirements are met. A web app may be configured to apply to specific file types within a multitype resource (such as a composite resource). This configuration should indicate whether the web app should appear on the resource level open with menu (making the menu entry only appear if there is a file of the specified file type in the resource) or at the file type level (making the app appear on a file level open with menu for a specific file type)."

1) The initial implementation I demo'd 2 weeks ago allows WebApp owner/editor to associate a webapp with one or more FileTypes. If being associated with more than one FileTypes, the relationship is "OR", which means this webapp can work separately with any of the selected FileTypes. Not to support "AND" or combination usecases, e.g a WebApp only works with a composite resource that has 2 raster filetypes + 1 feature filetypes). Are we on the same page?

2) With that said, since webapp only works with one instance of a FileType at a time, should the WebApp only show on filetype-level openwith list?

**Comment by dtarb** [comment]
@zhiyuli we are on the same page on point 1.  On point 2 it may be that a web app can work at resource level as long as there are certain file types present, perhaps more than 1.  An example is multiple netCDF files.  Open with from the resource level may want to point to a URL such as http://hyrax.hydroshare.org/opendap/31acb0c0a66f4de08f541077bdf3e1d2/data/contents/contents.html that provides an entry into the top level resource contents that would may have folders for multiple files.  But open with from the file level may go down to a file such as http://hyrax.hydroshare.org/opendap/36ca923a48b94114a9e1fe9651c7dc9e/data/contents/SWE_time/SWE_time.nc.html

I am suggesting that the web app writer should be able to specify separate URL launching patterns for resource level and file level entry if that makes sense for their particular application.

**Comment by dtarb** [comment]
@horsburgh the idea of this being editable from the open with list itself is appealing.  Speaking to @zhiyuli he suggested it be a second UI phase of the work, that could use the same backend functionality as what he could implement more easily for a "Show on Open With" check box.  

The comment about CUAHSI approved misses the point, because CUAHSI approved apps will always appear on Open With (step 2 in the business logic).  Step 3 is to enable a user to add an app that they want to use that is not CUAHSI approved, which is the point of much of this.  If we decided to not allow open with ever for non CUAHSI approved apps this would be simpler, but that negates the scalability of the app resource type that @danames team developed, and I think limits uses.

**Comment by zhiyuli** [comment]
@pkdash @alvacouch @Maurier @dtarb 

I reviewed "hs_lable" code and got the following findings (correct me please):

1) hs_lable supports 3 functionalities:
a) "Add to my resource": user actively discovers a public/published resource over which he has NO explicit permission, and then adds this resource to his "My Resource Page"
b) "Mark as Favorite": User marks a resource on his "My Resource Page" as favorite
c) "Label a Resource": User first creates labels, and then labels resources on his "My Resource Page"

2) (a) and (b) store user-resource-flag relations in the same table UserResourceFlags, where flag value has only 2 choices: 'MINE' for (a), and 'FAVORAITE' for (b)
    (c) stores user-resource-label relations in table UserResourceLabels, where label value can be any user-defined text.

I think I have 3 options here:
1) reuse current view functions and table UserResourceFlags, just to add a third flag choice-- "OPEN_WITH_APPS". Will this new flag choice make migration difficult?
2) reuse current view functions and table UserResourceLabels, just hard code a label value -
 "MyOpenWIthApps", in which case this label value is reserved for internal user only (User should not create this label any longer)
3) create a new Table to store user-app relation and add new necessary view functions

What do you think? Thanks

**Comment by dtarb** [comment]
My (weak) intuition is that option 1 seems best, but I would really prefer to defer to @pkdash @alvacouch @Maurier who know this code.

**Comment by zhiyuli** [comment]
@dtarb @horsburgh @danames 
This feature has been deployed on riverdev.hydroshare.org for testing.

**Comment by dtarb** [comment]
I believe this issue has been addressed in current deployments.  Closing.  @zhiyuli if there is remaining functionality needed here, I suggest you open a new issue with just that functionality.
